ios/UARCTModule/UARCTEventEmitter.m

```diff
@@ -36,37 +40,50 @@ - (instancetype)init {
     self = [super init];
     if (self) {
         self.pendingEvents = [NSMutableArray array];
+        self.knownListeners = [NSMutableSet set];
     }
 
     return self;
 }
 
-- (BOOL)sendEventWithName:(NSString *)eventName body:(id)body {
-    if (self.bridge && self.isObserving) {
+- (void)sendEventWithName:(NSString *)eventName body:(id)body {
+    if (self.bridge && self.isObserving && [self.knownListeners containsObject:eventName]) {
         [self.bridge enqueueJSCall:@"RCTDeviceEventEmitter"
                             method:@"emit"
                               args:body ? @[eventName, body] : @[eventName]
                         completion:NULL];
 
-        return YES;
+    } else {
+        @synchronized(self.pendingEvents) {
+            [self.pendingEvents addObject:@{ UARCTEventNameKey: eventName, UARCTEventBodyKey: body}];
+        }
     }
-
-    return NO;
 }
 
 - (void)addListener:(NSString *)eventName {
     self.listenerCount++;
-    if (self.listenerCount > 0) {
-        for (NSDictionary *event in self.pendingEvents) {
-            [self sendEventWithName:event[@"name"] body:event[@"body"]];
+
+    @synchronized(self.pendingEvents) {
+        for (id event in [self.pendingEvents copy]) {
+            if ([event[UARCTEventNameKey] isEqualToString:eventName]) {
+                [self sendEventWithName:event[UARCTEventNameKey] body:event[UARCTEventBodyKey]];
+                [self.pendingEvents removeObject:event];
+            }
         }
```

Comment:
Shoot. Now an event arriving during this loop will be added to `pendingEvents`, and never sent to the app. @marc-scig might be referring to the same thing. I think the entire `sendEventWithName:` and `addListener:` methods need to be synchronized. Don't do anything in either if the other one is executing. Maybe that can be done with `@synchronized(self.knownListeners)` at the start of each method?